### PR TITLE
Add validation to Parameter.assign() to increase robustness of transforms

### PR DESCRIPTION
**Pull request by st--**
In GPflow2, `Parameter` behaves differently from GPflow1, as noted in #1197.
Specifically, when using a positive transform, initializing or assigning the value 0.0 is perfectly fine in the forward pass: `p = gpflow.Parameter(0.0, transform=gpflow.utilities.positive())`, and it looks like nothing is wrong; the constrained representation is 0.0 as expected. (In GPflow1, it would implicitly force the value to be the lower bound of 1e-6.) However, the *unconstrained* representation is then `-inf`, which leads to `nan`s when computing gradients in the backward pass. This is error-prone and hard to debug. To make this issue more obvious, this PR adds an error check.

### CODE REVIEW

File: gpflow/base.py [hunk 1]
```diff
@@ -112,6 +113,13 @@ def trainable(self, flag: Union[bool, int]):
     def initial_value(self):
         return self._unconstrained.initial_value
 
+    def validate_unconstrained_value(self, value: tf.Tensor, dtype: DType) -> tf.Tensor:
+        value = _cast_to_dtype(value, dtype)
+        unconstrained_value = _to_unconstrained(value, self.transform)
+        return tf.debugging.assert_all_finite(unconstrained_value,
+                message=f"gpflow.Parameter.validate_unconstrained_value")
```
Comment (awav) [inline gpflow/base.py]: Could you change the message to something more descriptive?

File: gpflow/base.py [hunk 2]
```diff
@@ -225,9 +232,7 @@ def _run_op(a, *args, **kwargs):
 tf.register_tensor_conversion_function(Parameter, lambda x, *args, **kwds: x.read_value())
 
 
-def _verified_value(value: VariableData, dtype: Optional[DType] = None) -> np.ndarray:
-    if isinstance(value, tf.Variable):
-        return value
+def _cast_to_dtype(value: VariableData, dtype: Optional[DType] = None) -> tf.Tensor:
```
Comment (awav) [inline gpflow/base.py]:
```suggestion
def _cast_to_default_dtype(value: VariableData, dtype: Optional[DType] = None) -> tf.Tensor:
```

File: gpflow/base.py [hunk 3]
```diff
@@ -51,15 +51,16 @@ def __init__(self,
         """
         super().__init__()
 
-        value = _verified_value(value, dtype)
+        self._transform = transform
```
Comment (awav) [inline gpflow/base.py]: Could you move it back to the end? The reason is that attribute assign should happen after all potential failure cases.

File: gpflow/base.py (hunk 1 above)
Comment (st--) [reply gpflow/base.py]: What would you suggest?

File: gpflow/base.py (hunk 2 above)
Comment (st--) [reply gpflow/base.py]: It's not the default_dtype if the `dtype` argument is actually passed in. We could add a docstring, if you want, though it's an internal (module-private) function and very short so I thought that'd be overkill.

File: gpflow/base.py (hunk 3 above)
Comment (st--) [reply gpflow/base.py]: No I can't, I moved it to the front so that it's ready for validate_unconstrained_value() before the Variable construction.

**Comment by st--** [comment]
This change means it is no longer possible to actually assign `inf` or `nan` to a gpflow.Parameter. Are there any use-cases where we would actually want to do this ?

**Comment by jameshensman** [comment]
This all looks very sensible, and the test look like it covers the case you're interested in. 

I think the only thing to discuss is the behaviour. When doing `Parameter(0.0, positive)`:

1) the old behaviour is to get `1e-6` implicitly
2) the current (pre-this PR) is to get a NaN (bad! must fix.)
3) this PR would give you an exception
4) I wonder if it would be nicer to get a Warning and a change to `1e-6`?

File: gpflow/base.py (hunk 1 above)
Comment (awav) [reply gpflow/base.py]: "Unconstrained value of passed value has `NaN` or `Inf` and cannot be assigned"

**Comment by markvdw** [comment]
I agree with @st-- that an error would be the cleanest thing to do.

It's rare to want to initialise your parameters at the total edge of the boundary condition anyway, and if you are, it's probably best to be aware that you're not really able to do so!

**Comment by markvdw** [comment]
Although, just a thought... A related issue sometimes happens when the optimisation pushes the value to the utmost extreme. It would be annoying to optimise your model for ages, only for it to fail at the stage where the parameters are assigned.

I think this is the only use-case where I think that an error may not be the most appropriate thing? It would be nice if the result of the optimisation wouldn't be lost somehow...

**Comment by st--** [comment]
@markvdw I don't think this would be an issue as the optimizers only work in unconstrained space and directly assign to the actual (unconstrained_)variable, not the (constrained) parameter. So the _validate_unconstrained_value won't be called, and no assert error can be triggered.

**Comment by st--** [comment]
(we can also change it up again if it turns out to be an issue for anyone)

**Comment by markvdw** [comment]
Ah right, if that's the case, then that's great! Things change so quickly... I think in the past `assign()` was called on finishing the optimisation...

**Comment by st--** [comment]
Is there anything that needs doing on this PR then? If not please approve so we can merge it in!

File: gpflow/base.py (hunk 3 above)
Comment (awav) [reply gpflow/base.py]: Ok, then this is a separate issue, dependencies in the constructor calls and existing functions.

File: gpflow/base.py [hunk 4]
```diff
@@ -112,6 +113,14 @@ def trainable(self, flag: Union[bool, int]):
     def initial_value(self):
         return self._unconstrained.initial_value
 
+    def validate_unconstrained_value(self, value: tf.Tensor, dtype: DType) -> tf.Tensor:
+        value = _cast_to_dtype(value, dtype)
+        unconstrained_value = _to_unconstrained(value, self.transform)
+        message = "gpflow.Parameter: unconstrained value of passed value " \
+                  "has NaN or Inf and cannot be assigned."
+        return tf.debugging.assert_all_finite(unconstrained_value, message=message)
+
```
Comment (awav) [inline gpflow/base.py]:
```suggestion
```

**Pull request merged**